lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -349,6 +352,77 @@ def _promtail_error(self, event):
     )
 ```
 
+## LogForwarder Library Usage
+
+Let's say that we have a charm's workload that writes logs to the standard output (stdout),
+and we need to send those logs to a workload implementing the `loki_push_api` interface,
+such as `Loki` or `Grafana Agent`. To know how to reach a Loki instance, a charm would
+typically use the `loki_push_api` interface.
+
+Use the `LogForwarder` class by instantiating it in the `__init__` method of the charm.
+The easiest way to pass a Loki endpoint to the forwarder is to provide it with the name of
+an endpoint which uses the `loki_push_api` interface. The forwarder will automatically extract
+from there all the data it needs.
+
+1. To use the LogForwarder, the first you need to do is instantiate in in your charm's
+   constructor:
+
+   ```python
+   from charms.loki_k8s.v1.loki_push_api import LogForwarder
+
+   ...
+
+      def __init__(self, *args):
+          ...
+          self._log_forwarder = LogForwarder(
+              self,
+              relation_name="logging"  # optional, defaults to `logging`
+          )
+   ```
+
+   The `LogForwarder` by default will observe relation events on the `logging` endpoint and
+   enable/disable log forwarding automatically.
+   Next, modify the `metadata.yaml` file to add:
+
+   - The `log-forwarding` relation in the `requires` section:
+     ```yaml
+     requires:
+       logging:
+         interface: loki_push_api
+         optional: true
+     ```
+
+2. If you don't want to relate your charm to another implementing the `loki_push_api` interface,
+   you need to provide two things: the `relation_name` of the relation containing the endpoint(s),
+   and a `loki_endpoints_getter` function that extracts the endpoint(s) from that relation.
+   For example, let's say the charm gets the endpoint(s) from the `foo` relation:
+
+   ```python
+   from charms.loki_k8s.v1.loki_push_api import LogForwarder
+
+   ...
+
+     def __init__(self, *args):
+         ...
+         self._log_forwarder = LogForwarder(
+             self,
+             relation_name="foo",
+             loki_endpoints_getter=self.get_loki_endpoints("foo"),
+         )
+
+     def get_loki_endpoints(self, relation_name: str) -> List[str]:
+         endpoints = []
+         for relation in self.model.relations[relation_name]:
+             endpoint = some_way_to_retrieve_the_endpoint()
+             endpoints.append(endpoint)
+
+         return endpoints
```

Comment:
is there anything we should be clear about what kind of endpoints these can be? 
List[str] is a bit broad.
Do they need to be fully resolvable URLS? Hostnames only? `hostname:port`?

I'm thinking we need something like:
 "resolvable url of a `loki-push-api`-compliant endpoint"